kroxylicious-api/src/main/java/io/kroxylicious/proxy/service/BaseContributor.java

```diff
@@ -17,47 +18,47 @@
  *
  * @param <T> the service type
  */
-public abstract class BaseContributor<T> implements Contributor<T> {
+public abstract class BaseContributor<T, S extends ContributorContext> implements Contributor<T, S> {
 
-    private final Map<String, InstanceBuilder<? extends BaseConfig, T>> shortNameToInstanceBuilder;
+    private final Map<String, InstanceBuilder<? extends BaseConfig, T, S>> shortNameToInstanceBuilder;
 
     /**
      * Constructs and configures the contributor using the supplied {@code builder}.
      * @param builder builder
      */
-    public BaseContributor(BaseContributorBuilder<T> builder) {
+    public BaseContributor(BaseContributorBuilder<T, S> builder) {
         shortNameToInstanceBuilder = builder.build();
     }
 
     @Override
     public Class<? extends BaseConfig> getConfigType(String shortName) {
-        InstanceBuilder<?, T> instanceBuilder = shortNameToInstanceBuilder.get(shortName);
+        InstanceBuilder<?, T, S> instanceBuilder = shortNameToInstanceBuilder.get(shortName);
         return instanceBuilder == null ? null : instanceBuilder.configClass;
     }
 
     @Override
-    public T getInstance(String shortName, BaseConfig config) {
-        InstanceBuilder<? extends BaseConfig, T> instanceBuilder = shortNameToInstanceBuilder.get(shortName);
-        return instanceBuilder == null ? null : instanceBuilder.construct(config);
+    public T getInstance(String shortName, BaseConfig config, S context) {
+        InstanceBuilder<? extends BaseConfig, T, S> instanceBuilder = shortNameToInstanceBuilder.get(shortName);
+        return instanceBuilder == null ? null : instanceBuilder.construct(context, config);
     }
 
-    private static class InstanceBuilder<T extends BaseConfig, L> {
+    private static class InstanceBuilder<T extends BaseConfig, L, S extends ContributorContext> {
 
         private final Class<T> configClass;
-        private final Function<T, L> instanceFunction;
+        private final BiFunction<S, T, L> instanceFunction;
 
-        InstanceBuilder(Class<T> configClass, Function<T, L> instanceFunction) {
+        InstanceBuilder(Class<T> configClass, BiFunction<S, T, L> instanceFunction) {
             this.configClass = configClass;
             this.instanceFunction = instanceFunction;
         }
 
-        L construct(BaseConfig config) {
+        L construct(S context, BaseConfig config) {
             if (config == null) {
                 // tests pass in a null config, which some instance functions can tolerate
-                return instanceFunction.apply(null);
+                return instanceFunction.apply(context, null);
             }
             else if (configClass.isAssignableFrom(config.getClass())) {
-                return instanceFunction.apply(configClass.cast(config));
+                return instanceFunction.apply(context, configClass.cast(config));
```

Comment:
I need to process the PR properly, but my initial thought was that the context was going to provide the config as well. 

So this would end up something like:


```suggestion
                return instanceFunction.apply(context.wrap(configClass.cast(config)));
```

I'm not sure that is helpful or useful just what I had imagined was going to happen

Reply:
Yeah that was the alternative I had a go at, changing the signature to `T getInstance(String shortName, ContributorContext<? extends BaseConfig> context);`

I was struggling a little with the generics in BaseContributor but maybe it's better to live with an unchecked cast somewhere if the API is better.

Reply:
Yeah it became a bit awkward, if you make the interface:

```
interface ContributorContext {
    BaseConfig getConfig()
}
```

We don't have a Config at the point where we want to construct a FilterContributorContext, so we have to do something ugly like return null until we could call a method like `context.wrap(config)`.

So it feels like the FilterContributorContext has a different scope, we can use the same FilterContributorContext to construct many filters in the chain. But the Config varies through the chain (each definition provides a new config).

So I think it may make more sense to have them as separate parameters. Though it would be nice if we didn't have this empty marker interface used by the micrometer and cluster network config provider contributors.

Another option would be to remove the interface entirely and just use a type parameter. The other contributors could be `Contributor<MicrometerHook, Void>`.

Reply:
Would something like this work?

```java
public class ContributorContextFactory<C extends BaseConfig> implements Function<C, ContributorContext> {
    @Override
    public ContributorContext apply(C baseConfig) {
        return new ContributorContext(baseConfig);
    }
}
```

So the InstanceBuilder has a `ContributorContextFactory` and it invokes that each time thus avoiding the need for a wrap with a null field

Reply:
Not sure, another problem I hit was that because we are working with a varying ContributorContext sub-type we would want a way to wrap each type. Like if I have a FilterContributorContext, I don't want to wrap it with a ContributorContext, when I need it to remain a FilterContributorContext.